@trpc/next-layout/createTRPCNextLayout.ts

```diff
@@ -53,19 +57,109 @@ export type DecoratedProcedureRecord<
     : never;
 }>;
 
+type CreateTRPCNextLayout<TRouter extends AnyRouter> = DecoratedProcedureRecord<
+  TRouter["_def"]["record"]
+> & {
+  dehydrate(): Promise<DehydratedState>;
+};
+
+function getQueryKey(path: string[], input: unknown) {
+  return input === undefined ? [path] : [path, input];
+}
+
 export function createTRPCNextLayout<TRouter extends AnyRouter>(
   opts: CreateTRPCNextLayoutOptions<TRouter>,
-): DecoratedProcedureRecord<TRouter["_def"]["record"]> {
-  return createRecursiveProxy((callOpts) => {
+): CreateTRPCNextLayout<TRouter> {
+  function getState() {
+    const requestStorage = getRequestStorage<{
+      _trpc: {
+        cache: Record<
+          string,
+          {
+            updatedAt: Date;
+            data: unknown;
+            queryKey: QueryKey;
+          }
+        >;
+        context: inferRouterContext<TRouter>;
+      };
+    }>();
+    requestStorage._trpc = requestStorage._trpc ?? {
+      cache: Object.create(null),
+      context: opts.createContext(),
+    };
+    return requestStorage._trpc;
+  }
+  const transformer = opts.transformer ?? {
+    serialize: (v) => v,
+    deserialize: (v) => v,
+  };
+  return createRecursiveProxy(async (callOpts) => {
     const path = [...callOpts.path];
-    path.pop();
-    return use(
-      (async function iife() {
-        const ctx = await opts.createContext();
-        const caller = opts.router.createCaller(ctx);
+    const lastPart = path.pop();
+    const state = getState();
+    const ctx = await state.context;
+
+    if (lastPart === "dehydrate" && path.length === 0) {
+      await state.context;
+
+      const dehydratedState: DehydratedState = {
+        queries: [],
+        mutations: [],
+      };
+
+      for (const obj of Object.values(state.cache)) {
+        dehydratedState.queries.push({
+          // Does this matter?
+          queryHash: JSON.stringify(obj.queryKey),
+          queryKey: obj.queryKey,
+          state: {
+            data: obj.data,
+            error: null,
+            dataUpdateCount: 0,
+            dataUpdatedAt: obj.updatedAt.getTime(),
+            errorUpdateCount: 0,
+            errorUpdatedAt: obj.updatedAt.getTime(),
+            fetchFailureCount: 0,
+            fetchFailureReason: null,
+            fetchMeta: null,
+            fetchStatus: "idle",
+            isInvalidated: false,
+            status: "success",
+          },
```

Comment:
cc @TkDodo 

FYI I couldn't create a `new QueryClient()` on the server because that made Next.js blow up, which means I also couldn't use the `dehydrate()`-helpers & had to do a lot of ugly hacks on my own

Reply:
If the errors are related to not having React functions available in RCS you could try importing the `QueryClient` from `@tanstack/query-core`.

Reply:
Its related to query client having a default context.  query-core is a good option 👍🏾